cub/agent/agent_reduce.cuh

```diff
@@ -27,357 +27,418 @@
+  WrappedInputIteratorT d_wrapped_in; ///< Wrapped input data to reduce
+  ReductionOp reduction_op;           ///< Binary reduction operator
+
+  //---------------------------------------------------------------------
+  // Utility
+  //---------------------------------------------------------------------
+
+  // Whether or not the input is aligned with the vector type (specialized for
+  // types we can vectorize)
+  template <typename Iterator>
+  static __device__ __forceinline__ bool
+  IsAligned(Iterator d_in, Int2Type<true> /*can_vectorize*/)
+  {
+    return (size_t(d_in) & (sizeof(VectorT) - 1)) == 0;
+  }
+
+  // Whether or not the input is aligned with the vector type (specialized for
+  // types we cannot vectorize)
+  template <typename Iterator>
+  static __device__ __forceinline__ bool
+  IsAligned(Iterator /*d_in*/, Int2Type<false> /*can_vectorize*/)
+  {
+    return false;
+  }
+
+  //---------------------------------------------------------------------
+  // Constructor
+  //---------------------------------------------------------------------
+
+  /**
+   * @brief Constructor
+   * @param temp_storage Reference to temp_storage
+   * @param d_in Input data to reduce
+   * @param reduction_op Binary reduction operator
+   */
+  __device__ __forceinline__ AgentReduce(TempStorage &temp_storage,
+                                         InputIteratorT d_in,
+                                         ReductionOp reduction_op)
+      : temp_storage(temp_storage.Alias())
+      , d_in(d_in)
+      , d_wrapped_in(d_in)
+      , reduction_op(reduction_op)
+  {}
+
+  //---------------------------------------------------------------------
+  // Tile consumption
+  //---------------------------------------------------------------------
+
+  /**
+   * @brief Consume a full tile of input (non-vectorized)
+   * @param block_offset The offset the tile to consume
+   * @param valid_items The number of valid items in the tile
+   * @param is_full_tile Whether or not this is a full tile
+   * @param can_vectorize Whether or not we can vectorize loads
+   */
+  template <int IS_FIRST_TILE>
+  __device__ __forceinline__ void ConsumeTile(AccumT &thread_aggregate,
+                                              OffsetT block_offset,
+                                              int /*valid_items*/,
+                                              Int2Type<true> /*is_full_tile*/,
+                                              Int2Type<false> /*can_vectorize*/)
+  {
+    AccumT items[ITEMS_PER_THREAD];
+
+    // Load items in striped fashion
+    LoadDirectStriped<BLOCK_THREADS>(threadIdx.x,
+                                     d_wrapped_in + block_offset,
+                                     items);
+
+    // Reduce items within each thread stripe
+    thread_aggregate =
+      (IS_FIRST_TILE)
+        ? internal::ThreadReduce(items, reduction_op)
+        : internal::ThreadReduce(items, reduction_op, thread_aggregate);
```

Comment:
My knowledge here is not yet up to date, so please excuse if this is stupid.

Why do we treat the first tile specially? From my understanding it should suffice to properly initialize `thread_aggregate` and treat all tiles equally

Reply:
That's what the first overload is doing. Do you think that having:
```cpp
if (IS_FIRST_TILE) { thread_aggregate = input[0]; }
```
before the `ThreadReduce` invokation is better in any way?

Reply:
I think it would simplify the login here which is beneficial

Reply:
I'm not sure what you mean by login. We can keep the code as it was written and investigate the change you are referring to later.